src/main/java/net/minecraftforge/event/entity/player/MerchantTradeEvent.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * MerchantTradeEvent is fired whenever an event involving merchant trades occurs. <br>
+ * If a method utilizes this {@link Event} as its parameter, the method will
+ * receive every child event of this class.<br>
+ * <br>
+ * {@link #merchant} contains the merchant with whom to barter.<br>
+ * {@link #trade} contains the recipe being traded.<br>
+ * {@link #left} contains the itemstack in the left side.<br>
+ * {@link #right} contains the itemstack in the right side.<br>
+ * <br>
+ * All children of this event are fired on the {@link MinecraftForge#EVENT_BUS}.
+ **/
+public class MerchantTradeEvent extends PlayerEvent
+{
+    private final IMerchant merchant;
+    private final MerchantRecipe trade;
+    private final ItemStack left;
+    private final ItemStack right;
+
+    public MerchantTradeEvent(@Nonnull IMerchant merchant, @Nonnull MerchantRecipe trade, @Nonnull ItemStack left, @Nonnull ItemStack right)
+    {
+        super(merchant.getCustomer());
+        this.merchant = merchant;
+        this.trade = trade;
+        this.left = left;
+        this.right = right;
+    }
+
+    @Nonnull
+    public IMerchant getMerchant()
+    {
+        return this.merchant;
+    }
+
+    @Nonnull
+    public MerchantRecipe getTrade()
+    {
+        return this.trade;
+    }
+
+    @Nonnull
+    public ItemStack getLeft()
+    {
+        return this.left.copy();
+    }
+
+    @Nonnull
+    public ItemStack getRight()
+    {
+        return this.right.copy();
+    }
+
+
+    /**
+     * MerchantTradeEvent.CanTrade is fired before the result of trade appears.
+     * <br>
+     * The event is fired during the {@link InventoryMerchant#resetRecipeAndSlots()} method invocation.<br>
+     * <br>
+     * This event is not {@link Cancelable}.<br>
+     * <br>
+     * This event has a result. {@link HasResult}<br>
+     * {@link Result#ALLOW} The result of trade appears.<br>
+     * {@link Result#DENY} The result of trade does not appear.<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+     **/
+    @HasResult
+    public static class CanTrade extends MerchantTradeEvent
+    {
+        public CanTrade(@Nonnull IMerchant merchant, @Nonnull MerchantRecipe trade, @Nonnull ItemStack first, @Nonnull ItemStack second)
+        {
+            super(merchant, trade, first, second);
+        }
+    }
+
+    /**
+     * MerchantTradeEvent.Done is fired after the player takes itemstack.
+     * <br>
+     * The event is fired during the {@link SlotMerchantResult#onTake(EntityPlayer, ItemStack)} method invocation.<br>
+     * <br>
+     * This event is {@link Cancelable}.<br>
+     * If this event is canceled, the itemstack being about to be sold by the player is not shrunk.<br>
+     * <br>
+     * This event does not have a result. {@link HasResult}<br>
+     * <br>
+     * This event is fired on the {@link MinecraftForge#EVENT_BUS}.<br>
+     **/
+    @Cancelable
+    public static class Done extends MerchantTradeEvent
```

Comment:
I would name this one `Transact`, specifying that this is fired after the recipe is already created and when the player do the actual transaction.